File: r_generator/generate_data.R


## Generate a dataset that contains a true submodel of 'ntrue' regressors

## The data matrix X is drawn from the standard normal distribution

## The coefficients are 0 or 1. There a 'ntrue' 1 coefficients randomly
##     selected out of the range 1..nreg

## The error vector is drawn from the normal distribution with zero mean and
## standard deviation 'sd'


## nobs: numeric, number of observation in the model
## nreg: numeric, total number of regressors
## ntrue: numeric, number of "true" regressors
## intercept: logical
## sd: numeric, standard deviation of the error term

M_PI = 3.14159265358979323846

model <- function(nobs = 80, nreg = 60, ntrue = 30 , intercept = TRUE, sd = 0.01)
{
    ## independent variables
    X <- rnorm(nobs * nreg,sd=0.01)

    X = to_6_digits(X)

    dim(X) <- c(nobs, nreg)

    
    ## coefficients
    coefs <- rep_len(0, length.out = nreg)

    ##coefs_real <- runif(nreg, min=-0.5, max=1.5)

    coefs = to_6_digits(coefs)

    true <- sample.int(nreg, size = ntrue)

    coefs[true] <- 1 ##coefs_real[true]

    ## error
    error <- rnorm(nobs, sd = sd)

    error = to_6_digits(error)


    ## dependent variable
    y <- cbind(intercept, X) %*% c(1, coefs) + error

    y = to_6_digits(y)

    model = list(y = y, X = X, nobs = nobs, nreg = nreg, ntrue= ntrue, intercept = intercept, true = true, error = error)
 
}

norm_vec <- function(x) (sum(x^2))
fitness_func <- function(n, k, RSS) ((n + n * log(2 * M_PI) + n * log(RSS / n)+ 2 * (k + 2))) #k+1 without intercept

to_6_digits <- function(x)
{
    for (i in seq_along(x)){
     x[i] = as.numeric(formatC(round(x[i], 6), digits = 6, format = "f"))
        ## x[i] = as.numeric(format(round(x[i], 5), nsmall = 5))
    }

    to_6_digits = x
}

inc <- function(x)
{
 eval.parent(substitute(x <- x + 1))
}

write_to_file <-function(filename)
{
    temp_model = model()

    setwd("./generated_data")
    file.create(filename)

    ##get number of lines and columns to be written to file
    nlines = as.numeric(temp_model[3])
    ncolumns = as.numeric(temp_model[4]) + 1
    ncolumns_model = as.numeric(temp_model[5])

    y = unlist(temp_model[1])
    x = unlist(temp_model[2])

    combine = matrix(c(y,x),nrow = nlines,ncol = ncolumns,byrow = FALSE)

    write.table(combine, file=filename, row.names=FALSE, col.names=TRUE)

    best = (unlist(temp_model[7]))

    #print(temp_model[7])
    print(best)

    e = unlist(temp_model[8])
    RSS = norm_vec(e)
    AIC = fitness_func(ncolumns, ncolumns_model, RSS)

    print(AIC)
}

filename <- commandArgs(trailingOnly = TRUE)
##filename = "Gdata0.txt"
write_to_file(filename)

File: python_script/generate_data.py
import numpy as np
import sys

## nobs: numeric, number of observation in the model
## nreg: numeric, total number of regressors
## ntrue: numeric, number of "true" regressors
## intercept: logical
## sd: numeric, standard deviation of the error term

nobs = 100
nreg = 30
ntrue = 15
sd = 0.01
intercept = True
M_PI = 3.14159265358979323846


def generate_data():

    # generate X
    X = np.random.normal(size=nobs*nreg, scale=sd)
    X = np.reshape(X, (nobs, nreg))

    # generate coefficients
    coeffs = np.zeros(nreg)
    coeffs[:ntrue] = 1
    np.random.shuffle(coeffs)

    true = [i+1 for i, x in enumerate(coeffs) if x == 1]

    # generate error
    error = np.random.normal(size=nobs, scale=sd)

    # add intercept
    intercept = np.ones(shape=(nobs,1), dtype=np.int8)

    temp_x =      np.concatenate((intercept, X), axis=1)
    temp_coeffs = np.concatenate((np.array([1]), coeffs), axis=0)
    y = np.matmul(temp_x, temp_coeffs) + error

    return [y, X, error, nobs, nreg, ntrue, true]


def write_to_file(filename):

    temp_model = generate_data()

    y = temp_model[0].copy()
    x = temp_model[1].copy()
    err = temp_model[2].copy()
    best = temp_model[6].copy()
    n = temp_model[4]
    k = temp_model[5]

    filepath = "./generated_data/"

    file = filepath + filename

    try:
        f = open(file, "w+")

        # write first line
        for i in range(temp_model[4]+1):
            f.write('column{} '.format(i+1))
        f.write("\n")

        # write rest of lines
        for i in range(temp_model[3]):
            f.write('{} '.format(y[i]))
            for j in range(temp_model[4]):
                f.write('{} '.format(x[i][j]))
            f.write("\n")

    except Exception as err:
        print("Write to file error: ",err)

    try:
        RSS = sum([x**2 for x in err])
        AIC = (n + n * np.log(2 * M_PI) + n * np.log(RSS / n) + 2 * (k + 2)) #k+1 without intercept

    except ZeroDivisionError as err2:
        print("Compute RSS/AIC error", err2)

    if len(best) > 0:
        return [best, AIC]


generate_data()
